### DISCUSSION TITLE
strapi-plugin-rest-cache/guide/provider/custom-provider

### DISCUSSION THREAD

**Discussion by giscus** [strapi-community/strapi-plugin-rest-cache#89] [Documentation]
# strapi-plugin-rest-cache/guide/provider/custom-provider

Speed-up HTTP requests with LRU cache

https://strapi-community.github.io/strapi-plugin-rest-cache/guide/provider/custom-provider.html

**Comment by talkyyo** [comment]
Followed the guide exactly for the custom provider and get this error:

Error: Could not load REST Cache provider "../custom-rest-cache-provider/index.js". You may need to install a provider plugin "yarn add strapi-provider-rest-cache-../custom-rest-cache-provider/index.js".

**Reply by mtrofimiuk** [reply]
This is because this fragment in `bootstrap.js`:

```
  let modulePath;
  try {
    modulePath = require.resolve(`strapi-provider-rest-cache-${providerName}`);
  } catch (error) {
    if (error.code === 'MODULE_NOT_FOUND') {
      modulePath = providerName;
    } else {
      throw error;
    }
  }

  try {
    // eslint-disable-next-line
    provider = require(modulePath);
  } catch (err) {
    throw new Error(
      `Could not load REST Cache provider "${providerName}". You may need to install a provider plugin "yarn add strapi-provider-rest-cache-${providerName}".`
    );
  }
```

I do not understand why require in second try does not see local reference. I worked it around by creating my own package called `strapi-provider-rest-cache-memcached` and installed it with:
`yarn add file:./packages/strapi-provider-rest-cache-memcached`

**Reply by talkyyo** [reply]
Thanks this worked for me. I also noticed that if 2 different roles with 2 different accesses make the same call, the cache is still hit and returns a response. Do you know if there is some way to implement some sort of role-based authentication on the caching logic?

**Reply by mtrofimiuk** [reply]
This is how cache key is generated. It takes url and query string. You would need to provide own method and include a role to be included in a cache key. Than each role would have different cache keys.

https://github.com/strapi-community/strapi-plugin-rest-cache/blob/main/packages/strapi-plugin-rest-cache/server/utils/keys/generateCacheKey.js#L8

**Comment by princeisamaliya** [comment]
"rest-cache": {
    config: {
      provider: {
        name: "memory",
        options: {
          max: 32767,
          maxAge: 3,
          updateAgeOnGet: false,
        },
      },
      strategy: {
        debug: true,
        maxAge: 3600,
        updateAgeOnGet: false,
        contentTypes: [
          // list of Content-Types UID to cache
          "api::faq.faq",
          "api::news.news",
          "api::case-study.case-study",
          "api::case-study-category.case-study-category",
          "api::faq-category.faq-category",
          "api::page.page",
          "api::product-level-1.product-level-1",
          "api::product-level-2.product-level-2",
          "api::product-level-3.product-level-3",
          "api::product.product",
          "api::resource.resource",
          "api::resource-category.resource-category",
          "api::resource-sub-category.resource-sub-category",
          "api::footer.footer",
          "api::global.global",
          "api::header.header",
          "api::newsletter.newsletter",
        ],
      },
    },
  },

Above config seems not working for me on strapi 4 cloud hosting.
